massa-event-cache/src/event_cache.rs

```diff
@@ -0,0 +1,454 @@
+use std::cmp::max;
+// std
+use nom::AsBytes;
+use std::path::Path;
+// third-party
+use crate::ser_deser::{
+    SCOutputEventDeserializer, SCOutputEventDeserializerArgs, SCOutputEventSerializer,
+};
+use massa_models::execution::EventFilter;
+use massa_models::output_event::SCOutputEvent;
+use massa_serialization::{DeserializeError, Deserializer, Serializer};
+use rocksdb::{DBIterator, IteratorMode, WriteBatch, DB};
+use tracing::debug;
+
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const EVENT_DESER_ERROR: &str = "critical: event deserialization failed";
+
+/// Module key formatting macro
+#[macro_export]
+macro_rules! event_key {
+    ($event_slot:expr) => {
+        [&$event_slot.to_bytes()[..], &[MODULE_IDENT]].concat()
+    };
+}
+
+pub(crate) struct EventCache {
+    /// RocksDB database
+    db: DB,
+    /// How many entries are in the db. Count is initialized at creation time by iterating
+    /// over all the entries in the db then it is maintained in memory
+    entry_count: usize,
+    /// Maximum number of entries we want to keep in the db.
+    /// When this maximum is reached `snip_amount` entries are removed
+    max_entry_count: usize,
+    /// How many entries are removed when `entry_count` reaches `max_entry_count`
+    snip_amount: usize,
+    /// Event serializer
+    event_ser: SCOutputEventSerializer,
+    /// Event deserializer
+    event_deser: SCOutputEventDeserializer,
+}
+
+impl EventCache {
+    /// Create a new EventCache
+    pub fn new(path: &Path, max_entry_count: usize, snip_amount: usize, thread_count: u8) -> Self {
+        let db = DB::open_default(path).expect(OPEN_ERROR);
+        let entry_count = db.iterator(IteratorMode::Start).count();
```

Comment:
clear the DB at start, set entry_count to 0 (see above)

Reply:
Done